### DISCUSSION TITLE
cannot resolve default property values of DataCollection type

### DISCUSSION THREAD

**Discussion by quyle92** [spatie/laravel-data#228] [Q&A]
Hi,
I have this DTO class with nested collection of `Decklist`:
```
class CharacterClassData extends Data
{
    public function __construct(
        /** @var DataCollection<Decklist> */
        public DataCollection $decklist,
    ) {
    }
}

class Decklist extends Data
{
    public function __construct(
        public ?string $action_id = "",
        public ?int $quantity = 0,
        public ?bool $innate = false,
    ) {
    }
}
```
I am trying to convert current model instance to a data object by using WithData trait:
```
class CharacterClass extends Model{
    use WithData;
    
    protected $dataClass = CharacterClassData::class;
}
```
but when using `getData()` method, I cannot get default values defined in `Decklist` class, i.e calling `$character_class->getData()` only returns empty array.
But I want it to return default values of defined in `Decklist` class like so:
```
 [[
    'action_id' => "",
    'quantity' => null,
    'innate' => null,
]]
```
How can I achieve this please?

**Comment by Nir-An** [comment]
Hi @quyle92 

I think that is not theoretical dto, we talking about dto with the context and data from the model!

so the default values should come from the model himself.

https://laravel.com/docs/9.x/eloquent#default-attribute-values

in MVC aspects you don't want more than one source to get the values.
In the end, a dto is only a kind of an object :+1:
the logic and default values leave to model to be changed on them.

**Reply by quyle92** [reply]
Hi @Nir-An ,
Thanks for your answer... But the issue is if the property's data type is of nested data object, default values can be retrieved, i.e. if I edit the above code from:
```
public function __construct(
        /** @var DataCollection<Decklist> */
        public DataCollection $decklist,
    ) {
    }
```
to:
```
public function __construct(
        public Decklist $decklist,
    ) {
    }
```
Then I can get the default `$decklist` values defined in `Decklist` class without any issue.

Also, the `$attributes` is used when you want to initialize a new instance from a model, i.e
`$character_class = new CharacterClass;`
then only in this case, default value that was defined in `$attributes` property would take effect.

**Comment by Nir-An** [comment]
Hi @quyle92 

Can you add please the imputation of the getData method? 

Thanks for your explanation! :+1:

**Reply by quyle92** [reply]
imputation of the `getData()`? I am not sure what you mean?

**Comment by rubenvanassche** [comment]
What's the value of `$character_class`? Is it an empty model or one that exists? Actually that probably doesn't matter since the data object will be constructed from the values of the model and ignoring the default values since they're defined on the model.

**Reply by quyle92** [reply]
> What's the value of `$character_class`? Is it an empty model or one that exists? Actually that probably doesn't matter since the data object will be constructed from the values of the model and ignoring the default values since they're defined on the model.

`$character_class` is a Model instance, but when the value of `$decklist` property is empty, instead of empty array, I want it to be able to return default values (which I've already defined in `Decklist` data class) like so:
```
 [[
    'action_id' => "",
    'quantity' => null,
    'innate' => null,
]]
```

**Reply by rubenvanassche** [reply]
The package cannot do this automatically do this for you, though I would take a look at custom creation methods. You probably could create one for the model and already add some default data there.
